core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java

```diff
@@ -0,0 +1,250 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.sort.io;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.spark.SparkConf;
+import org.apache.spark.api.shuffle.ShuffleMapOutputWriter;
+import org.apache.spark.api.shuffle.ShufflePartitionWriter;
+import org.apache.spark.shuffle.ShuffleWriteMetricsReporter;
+import org.apache.spark.internal.config.package$;
+import org.apache.spark.shuffle.IndexShuffleBlockResolver;
+import org.apache.spark.storage.TimeTrackingOutputStream;
+import org.apache.spark.util.Utils;
+
+public class DefaultShuffleMapOutputWriter implements ShuffleMapOutputWriter {
+
+  private static final Logger log =
+    LoggerFactory.getLogger(DefaultShuffleMapOutputWriter.class);
+
+  private final int shuffleId;
+  private final int mapId;
+  private final ShuffleWriteMetricsReporter metrics;
+  private final IndexShuffleBlockResolver blockResolver;
+  private final long[] partitionLengths;
+  private final int bufferSize;
+  private int currPartitionId = 0;
+
+  private final File outputFile;
+  private final File outputTempFile;
+  private FileOutputStream outputFileStream;
+  private FileChannel outputFileChannel;
+  private TimeTrackingOutputStream ts;
+  private BufferedOutputStream outputBufferedFileStream;
+
+  public DefaultShuffleMapOutputWriter(
+      int shuffleId,
+      int mapId,
+      int numPartitions,
+      ShuffleWriteMetricsReporter metrics,
+      IndexShuffleBlockResolver blockResolver,
+      SparkConf sparkConf) {
+    this.shuffleId = shuffleId;
+    this.mapId = mapId;
+    this.metrics = metrics;
+    this.blockResolver = blockResolver;
+    this.bufferSize =
+      (int) (long) sparkConf.get(
+        package$.MODULE$.SHUFFLE_UNSAFE_FILE_OUTPUT_BUFFER_SIZE()) * 1024;
+    this.partitionLengths = new long[numPartitions];
+    this.outputFile = blockResolver.getDataFile(shuffleId, mapId);
+    this.outputTempFile = Utils.tempFileWith(outputFile);
+  }
+
+  @Override
+  public ShufflePartitionWriter getNextPartitionWriter() throws IOException {
+    initStream();
+    initChannel();
+    return new DefaultShufflePartitionWriter(currPartitionId++);
+  }
+
+  @Override
+  public void commitAllPartitions() throws IOException {
+    cleanUp();
+    blockResolver.writeIndexFileAndCommit(shuffleId, mapId, partitionLengths, outputTempFile);
+    if (!outputFile.exists()) {
+      if (!outputFile.getParentFile().isDirectory() && !outputFile.getParentFile().mkdirs()) {
+        throw new IOException(
+          String.format(
+            "Failed to create shuffle file directory at %s.",
+            outputFile.getParentFile().getAbsolutePath()));
+      }
+      if (!outputFile.isFile() && !outputFile.createNewFile()) {
```

Comment:
Wait, i'm confused by this block. You're checking whether the output file is there, and then trying to create it? Shouldn't you be throwing if it's not there regardless of whether you can create it or not because when `blockResolver.writeIndexFileAndCommit()` is done, the file should be there? But let me know if I'm misunderstanding this.

Reply:
If the committer didn't create the file, we want to try to create an empty one by default, I think. But it's strange to me that shuffles can write empty files. Was this the case before our plugin refactor?

Reply:
Oh i see... well, looking at this code, I don't think it actually writes an empty file if there are no records: https://github.com/bloomberg/apache-spark-on-k8s/pull/6/files#diff-8b6b7a5dadc0d8e97307d0f8e8378d8fL126